modules/core/04-channel/keeper/upgrade.go

```diff
@@ -89,3 +98,41 @@ func (k Keeper) WriteUpgradeInitChannel(
 
 	emitChannelUpgradeInitEvent(ctx, portID, channelID, upgradeSequence, channelUpgrade)
 }
+
+// RestoreChannel restores the given channel to the state prior to upgrade.
+func (k Keeper) RestoreChannel(ctx sdk.Context, portID, channelID string, upgradeSequence uint64) error {
+	_, code, _ := errorsmod.ABCIInfo(types.ErrChannelRestored, false) // discard non-determinstic codespace and log values
+
+	errorReceipt := types.ErrorReceipt{
+		Sequence: upgradeSequence,
+		Error:    fmt.Sprintf("ABCI code: %d: %s", code, restoreErrorString),
+	}
+
+	k.SetUpgradeErrorReceipt(ctx, portID, channelID, errorReceipt)
+
+	channel, found := k.GetUpgradeRestoreChannel(ctx, portID, channelID)
+	if !found {
+		return errorsmod.Wrapf(types.ErrRestoreChannelNotFound, "channel-id: %s", channelID)
+	}
+
+	k.SetChannel(ctx, portID, channelID, channel)
+	k.DeleteUpgradeRestoreChannel(ctx, portID, channelID)
+	k.DeleteUpgradeTimeout(ctx, portID, channelID)
+
+	module, _, err := k.LookupModuleByChannel(ctx, portID, channelID)
+	if err != nil {
+		return errorsmod.Wrap(err, "could not retrieve module from port-id")
+	}
+
+	portKeeper, ok := k.portKeeper.(*portkeeper.Keeper)
+	if !ok {
+		panic("todo: handle this situation")
+	}
```

Comment:
would love some input about this. A few things were attempted.

1. Add an entry to expected keepers. - this created a circular import between 04-channel and 05-port
2. Add IBCModule to exported package. - this created a circular import between exported and 04-channel as the interface imports types from channel.

Casting directly to `portkeeper.Keeper` defeats the purpose of an expected keepers interface. Let me know if there is a nice way to work around this without introducing more packages.

Reply:
Maybe if we can [execute the callback at the message server layer](https://github.com/cosmos/ibc-go/pull/3352/files#r1151544321), then we don't have this problem? :)

Reply:
Should IBCModule be pulled out into its own separate package?

Maybe with one day giving it a separate go.mod

Reply:
Definitely worth considering, but I think its likely part of a much larger body of work and related to https://github.com/cosmos/ibc-go/issues/2471